middle_end/flambda/types/env/aliases.ml

```diff
@@ -289,127 +418,195 @@ let add_alias_between_canonical_elements t ~canonical_element ~to_be_demoted =
-      t.aliases_of_canonical_elements
-      |> Simple.Map.remove to_be_demoted
-      |> Simple.Map.add (* replace *) canonical_element aliases
+    let aliases_of_canonical_names, aliases_of_consts =
+      Simple.pattern_match canonical_element
+        ~name:(fun name ->
+          Name.Map.add (* replace *) name aliases aliases_of_canonical_names,
+          t.aliases_of_consts)
+        ~const:(fun const ->
+          aliases_of_canonical_names,
+          Const.Map.add (* replace *) const aliases t.aliases_of_consts)
     in
     { canonical_elements;
-      aliases_of_canonical_elements;
+      aliases_of_canonical_names;
+      aliases_of_consts;
       binding_times_and_modes = t.binding_times_and_modes;
     }
 
-type to_be_demoted = {
-  canonical_element : Simple.t;
-  to_be_demoted : Simple.t;
-}
+type to_be_demoted = Demote_canonical_element1 | Demote_canonical_element2
 
 let choose_canonical_element_to_be_demoted t ~canonical_element1
       ~canonical_element2 =
   if defined_earlier t canonical_element1 ~than:canonical_element2
-  then
-    { canonical_element = canonical_element1;
-      to_be_demoted = canonical_element2;
-    }
-  else
-    { canonical_element = canonical_element2;
-      to_be_demoted = canonical_element1;
-    }
+  then Demote_canonical_element2 else Demote_canonical_element1
 
 (* CR mshinwell: add submodule *)
 type add_result = {
   t : t;
   canonical_element : Simple.t;
-  alias_of : Simple.t;
+  alias_of_demoted_element : Simple.t;
 }
 
-let invariant_add_result ~original_t { canonical_element; alias_of; t; } =
+let invariant_add_result ~original_t { canonical_element; alias_of_demoted_element; t; } =
   if !Clflags.flambda_invariant_checks then begin
     invariant t;
-    if not (Simple.equal canonical_element alias_of) then begin
-      if not (defined_earlier t canonical_element ~than:alias_of) then begin
-        Misc.fatal_errorf "Canonical element %a should be defined earlier \
-            than %a after alias addition.@ Original alias tracker:@ %a@ \
+    if not (defined_earlier t canonical_element ~than:alias_of_demoted_element) then begin
+      Misc.fatal_errorf "Canonical element %a should be defined earlier \
+          than %a after alias addition.@ Original alias tracker:@ %a@ \
+          Resulting alias tracker:@ %a"
+        Simple.print canonical_element
+        Simple.print alias_of_demoted_element
+        print original_t
+        print t
+    end;
+    match canonical t alias_of_demoted_element with
+    | Is_canonical _ ->
+        Misc.fatal_errorf "Alias %a must not be must not be canonical \
+            anymore.@ \
+            Original alias tracker:@ %a@ \
             Resulting alias tracker:@ %a"
-          Simple.print canonical_element
-          Simple.print alias_of
+          Simple.print alias_of_demoted_element
           print original_t
           print t
-      end
-    end
+    | Alias_of_canonical _ -> ()
   end
 
+(*
```

Comment:
Please remove this debug code

Reply:
Yeah, it's a bit much. Wish we had a systematic means of debug logging.